feat: add BoxBuild bind

color-dev opened the pull request:
### Summary
Adds a new 'boxbuild' bind to the game.

### Rationale
Improves ergonomics for players by allowing boxbuild to be triggered via keybinds using a widget.

When you press boxbuild intended for mouse button, it clicks and does the filled rectangle action.
To build rectangles of 4 windmills with one click of the boxbuild keybind.
Or you can also click with mouse, press boxbuild, and it changes the mode instead of pressing alt.

### Testing
Verified locally that the bind triggers correctly with F14. BoxBuild widget calls Spring.EmulateMousePress and Spring.EmulateMouseRelease and command "build_state box" and "build_state box r" to execute the boxbuild action.

`uikeys.txt`:
bind            Any+F14  boxbuild
bind              shift  build_state line
bind         shift+ctrl  build_state circle
bind          shift+alt  build_state box
bind               ctrl  build_state axis_lock
bind     shift+ctrl+alt  build_state hollow_box

In-game command "/guihandler legacybuildui" toggles legacy and unhardcoded behavior
build behavior matches legacy

### Notes / Dependencies
This PR builds on top of https://github.com/beyond-all-reason/RecoilEngine/pull/2623
This feature requires a corresponding widget change in the **BAR main repo** for full functionality.  
- `Line-Build Hotkey Ignore` widget must whitelist the keybind.
- `Line-Build Hotkey Ignore` can be temporarily disabled until a whitelist update is merged.
- `Mouse Buildspacing` widget currently hardcodes MOUSE4/MOUSE5 for buildspacing changes.
- `Mouse Buildspacing` should be disabled you wish to use MOUSE4/MOUSE5 with the uikeys system.
- The earlier MouseHandler fixes for missed inputs using mouse4/5 binds have been split into a separate PR:
https://github.com/beyond-all-reason/RecoilEngine/pull/2613

### Status (last updated: Nov 8)
- MouseActionExecutor may be a better option than Spring.EmulateMousePress/Release

color-dev committed:
feat: add BoxBuild bind

color-dev committed:
fix: handle Mouse4/5 binding in GuiHandler. Corrected lastOpts refactor

color-dev committed:
fix: Mouse4/5 binding & lastOpts refactor

color-dev committed:
fix: remove default binding

color-dev committed:
Remove MouseHandler changes (moved to separate PR)

color-dev commented on rts/Game/UI/MouseHandler.cpp:
```diff
@@ -334,6 +334,24 @@ void CMouseHandler::MousePress(int x, int y, int button)
 
 	pressedBitMask |= 1 << button;
 
+	const bool isXButton = (button == SDL_BUTTON_X1 || button == SDL_BUTTON_X2);
```
Thanks for pointing that out. I’ve moved the MouseHandler changes into a separate PR (#2613)

badosu commented:
I already provided some feedback privately, but for the record and transparency sake I post it here (with a few adjustments):

It seems to me this does not truly unhardcode the modifier for setting up the "queue" mode: a state where the engine captures mouse input to provide the UX for: multiple disjoint units, line, or square blueprints.

As it is, one indeed could bind to `boxbuild` and have this mode enabled, e.g.: `bind ctrl+b boxbuild`. However there are a few issues with the current implementation:

- Most importantly: the engine hardcoded binding is still present: so ctrl, alt and shift _still_ perform the same job as current. This does not enable games to use the bind and clear up these modifiers to do something else.
- The currently defined actions only option is to set the "square" build mode, without any further flexibility.

The current effort and policy from the engine is to avoid as much game-specific _ad hoc_ behavior as possibly. Allowing games to define as much behavior as they want without relying on engine quirks.

With this in mind here are my suggestions:

#### Make it truly unhardcoded

If you bound to something that is not shift or alt, shift or alt should not trigger it.

This means storing state on what modifiers are currently active, for example see how [Camera stores modifier state for unhardcoded camera move bindings](https://github.com/beyond-all-reason/RecoilEngine/blob/3f0f87ab6f40b8d3e32268fa45a5d5ad63ad824a/rts/Game/Camera.cpp#L722-L734) and [how this state is modified via actions](https://github.com/beyond-all-reason/RecoilEngine/blob/3f0f87ab6f40b8d3e32268fa45a5d5ad63ad824a/rts/Game/UnsyncedGameCommands.cpp#L753-L763).

Make it optin and retrocompatible with current hardcoded bindings via a configuration variable (default current). See for example how a boolean variable can be defined and used [here](https://github.com/beyond-all-reason/RecoilEngine/blob/3f0f87ab6f40b8d3e32268fa45a5d5ad63ad824a/rts/Game/UI/MiniMap.cpp#L60), [here](https://github.com/beyond-all-reason/RecoilEngine/blob/3f0f87ab6f40b8d3e32268fa45a5d5ad63ad824a/rts/Game/UI/MiniMap.cpp#L100) and hot reloaded [here](https://github.com/beyond-all-reason/RecoilEngine/blob/3f0f87ab6f40b8d3e32268fa45a5d5ad63ad824a/rts/Game/UI/MiniMap.cpp#L366).

This would look something like:

```cpp
// pseudo-codeish

enableNativeBuildUI = config(default = true) // fetch configuration

// store modifier state
// note modifiers are disjoint and independent below, this helps address the issue
// of remembering our previous state was, so for example when box is released line is still true
buildStateModifiers = {
  BUILD_STATE_LINE = false,
  BUILD_STATE_BOX = false
}

GetCurrentBuildState() {
  if (enableNativeBuildUI) {
    GetNativeBuildState()
  } else {
    GetModifierBuildState()
  }
}

GetNativeBuildState() {
  if (!shift)
    return BUILD_STATE_NONE
  
  return alt ? BUILD_STATE_BOX : BUILD_STATE_LINE
}

// quick lazy impl just for reference
GetModifierBuildState() {
  if buildStateModifiers[BUILD_STATE_BOX]
    return BUILD_STATE_BOX
  if buildStateModifiers[BUILD_STATE_LINE]
    return BUILD_STATE_LINE
  
  return BUILD_STATE_NONE
}

CommandAssignmentOrDrawing() {
  switch GetModifierBuildState():
    case BUILD_STATE_BOX:
      // do box stuff
    case BUILD_STATE_LINE:
      // do line stuff
    case BUILD_STATE_NONE:
      // do none stuff
}
```

#### Improve action handling and future proof it

See for example the [unhardcoded group action](https://github.com/beyond-all-reason/RecoilEngine/blob/3f0f87ab6f40b8d3e32268fa45a5d5ad63ad824a/rts/Game/UnsyncedGameCommands.cpp#L1259-L1269):

```
group [add|set|focus|...] [groupIndex]
```

The functional concern is grouped by action and dispatched by arguments. For our case, this is an idea:

```
build_state [line|box]
```

Note that with the configuration option we are allowing games to disable currently hardcoded engine build UI and allowing games to define their own logic by intercepting the `build_state` action.

color-dev committed:
Expose build_state box action; groundwork for unbinding Alt/Shift/Ctrl

- Preserves boxbuild functionality by toggling build_state box
- Currently emulates mouse press/release to reuse existing input path
- Unsure if this should remain engine-side or be moved to widget layer

color-dev committed:
LuaUnsyncedCtrl: add Spring.EmulateMousePress/Release for widget use

color-dev committed:
input/gui: fix modifier-only chords, unhardcode build_state modifiers

color-dev committed:
fix: modifier-only chord bindings integrated with action system
